app/flows/user_confirmation_flow.rb

```diff
@@ -0,0 +1,65 @@
+class UserConfirmationFlow
+  attr_reader :user, :match, :receiver
+
+  def initialize(user_token)
+    @user_token = user_token
+    @success = false
+    @must_cleanup = false
+  end
+
+  def perform
+    ActiveRecord::Base.transaction do
+      @user = user_from_token or rollback!("no user with such token")
+
+      rollback!("registrations closed", cleanup: true) unless Settings.registrations_enabled?
+
+      @receiver = receiver_from_match or rollback!("no receivers available", cleanup: true)
+
+      @success = true
+    end
+  rescue ActiveRecord::ActiveRecordError => e
+    Rails.logger.error(e.message)
+  ensure
+    cleanup if must_cleanup
+  end
+
+  def successful?
+    success
+  end
+
+  def user_deleted?
+    must_cleanup
+  end
+
+  private
+
+  attr_reader :user_token, :success, :must_cleanup
+
+  def user_from_token
+    user_confirmation = UserConfirmation.new(user_token)
+    user_confirmation.perform
+
+    user_confirmation.successful? ? user_confirmation.user : nil
+  end
+
+  def receiver_from_match
+    match_assignment = MatchAssignment.new(user)
+    match_assignment.perform
+
+    Settings.put(Settings::REGISTRATIONS_ENABLED, false) if User.confirmed.count == Receiver.count
+
+    match_assignment.successful? ? match_assignment.receiver : nil
+  end
+
+  def cleanup
+    user&.destroy
+
+    Settings.put(Settings::REGISTRATIONS_ENABLED, false)
```

Comment:
Here it's more difficult, and there is a specific case where this needs to run.

The cleanup should be done when you want to destroy the user and cancel all registrations on a NOT successful confirmation.

The registration disable in `receiver_from_match` is done on a successful context, where you disable after matching the total number of users with the receivers.

Reply:
👍 that's tricky to know that. Should we put it in a comment? Otherwise it seems to be very redundant and we might forget about it in the future. What do you think?

Reply:
I'll comment.